pkg/operator/status/status_controller.go

```diff
@@ -0,0 +1,404 @@
+	if err != nil {
+		log.Fatal(err)
+	}
+	platformType := platform.GetType(infraStatus)
+
+	r := newReconciler(mgr, platformType)
+
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &operatorv1.CloudCredential{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+
+	// Whenever a CredentialsRequest is modified, recalculate status
+	err = c.Watch(&source.Kind{Type: &credreqv1.CredentialsRequest{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: handler.ToRequestsFunc(alwaysReconcileCCOConfigObject),
+	})
+	if err != nil {
+		return err
+	}
+
+	// These functions are used to determine if an event for the given Secret should trigger a sync
+	// We are watching for:
+	// 	Future known secrets to set appropriate Upgradeable conditions.
+	//	Secrets with the CCO annotation on them (secrets created by CCO).
+	//	The root cloud cred secret in the kube-system namespace.
+	p := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return isWatchedSecret(platformType, e.MetaNew.GetNamespace(), e.MetaNew.GetName(), e.MetaNew.GetAnnotations())
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return isWatchedSecret(platformType, e.Meta.GetNamespace(), e.Meta.GetName(), e.Meta.GetAnnotations())
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return isWatchedSecret(platformType, e.Meta.GetNamespace(), e.Meta.GetName(), e.Meta.GetAnnotations())
+		},
+	}
+
+	// Whenever one of our watched Secrets is updated, recalculate status
+	err = c.Watch(&source.Kind{Type: &corev1.Secret{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: handler.ToRequestsFunc(alwaysReconcileCCOConfigObject),
+		},
+		p,
+	)
+
+	// TODO: Watch the Authentication object for changes as that can cause the OIDC controller to change status
+	// (but something as simple as being unable to Get() the Authentication can cause degraded and we will not
+	// be notified of that).
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileStatus{}
+
+// ReconcileStatus reconciles the entire status for this operator.
+type ReconcileStatus struct {
+	client.Client
+	Logger   log.FieldLogger
+	platform configv1.PlatformType
+}
+
+// Reconcile will ensure the ClusterOperator status conditions are updating by calling to each
+// registered domain-specific status handler. This function will ensure default conditions are set
+// if none of the handlers set a condition.
+func (r *ReconcileStatus) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	start := time.Now()
+
+	defer func() {
+		dur := time.Since(start)
+		metrics.MetricControllerReconcileTime.WithLabelValues(controllerName).Observe(dur.Seconds())
+	}()
+
+	err := SyncStatus(r.Client, r.Logger)
+	return reconcile.Result{}, err
+}
+
+// SyncStatus allows triggering a status sync from outside of this controller.
```

Comment:
I'm really not sure if this is a bad thing. Seems fairly elegant, aside from error handling and requeues.